core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -422,11 +425,17 @@ private[log] class Cleaner(val id: Int,
       for (entry <- messages.shallowIterator) {
         val size = MessageSet.entrySize(entry.message)
         stats.readMessage(size)
-        if (entry.message.compressionCodec == NoCompressionCodec) {
+        val sourceCodec: CompressionCodec = entry.message.compressionCodec
+        val targetCodec = BrokerCompressionCodec.getTargetCompressionCodec(brokerCompressionType, sourceCodec)
+        if (sourceCodec == NoCompressionCodec) {
           if (shouldRetainMessage(source, map, retainDeletes, entry)) {
             val convertedMessage = entry.message.toFormatVersion(messageFormatVersion)
-            ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
-            stats.recopyMessage(size)
+            if (targetCodec == NoCompressionCodec) { // same as sourceCodec
+              ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
+              stats.recopyMessage(size)
+            } else {
+              compressMessages(writeBuffer, targetCodec, messageFormatVersion, Seq(new MessageAndOffset(convertedMessage, entry.offset)))
```

Comment:
Yes, this doesn't seem like the right way to do it. We would probably want to create outer messages with a number of inner messages each before compressing. It may be worth checking with @junrao what he had in mind for this case.